Sources/CvPlayerAI.cpp

```diff
@@ -12852,9 +12852,58 @@ int CvPlayerAI::AI_unitPropertyValue(UnitTypes eUnit, PropertyTypes eProperty) c
 	return iValue;
 }
 
-/********************************************************************************/
-/* 	City Defenders						24.07.2010				Fuyu			*/
-/********************************************************************************/
+int CvPlayerAI::AI_unitBuildingValue(UnitTypes eUnit, CvArea* pArea) const
+{
+	FAssertMsg(eUnit != NO_UNIT, "Unit is not assigned a valid value");
+	FAssertMsg(eUnitAI != NO_UNITAI, "UnitAI is not assigned a valid value");
+
+	CvUnitInfo& kUnitInfo = GC.getUnitInfo(eUnit);
+
+	// Evaluate unit's ability to construct buildings
+	int iCount = 0;
+
+	if (kUnitInfo.getNumBuildings() > 0 && pArea != NULL && !isNPC())
+	{
+		int iBuildingValue;
+		int iLoop;
+		bool bCoastal = kUnitInfo.getDomainType() == DOMAIN_SEA;
+		int iMinOceanSize;
+		if (bCoastal)
+		{
+			iMinOceanSize = GC.getMIN_WATER_SIZE_FOR_OCEAN();
+		}
+
+		for (int iI = 0; iI < kUnitInfo.getNumBuildings(); iI++)
+		{
+			BuildingTypes eBuilding = (BuildingTypes)kUnitInfo.getBuildings(iI);
+
+			if (NO_BUILDING != eBuilding)
```

Comment:
Is it ever actually possible for this to happen? Surely if you are looping over the buildings of the unit then they will all be valid buildings. If this is the case then this check should be an `FAssertMsg` instead like `FAssertMsg(eBuilding != NO_BUILDING, "UnitInfo buildings must be valid");`.
This improves readability, and performance, and better expresses expectations of the data.

Reply:
This was a conversion artifact when removing the buildingclass reference where it WAS possible for eBuilding to refer to NO_BUILDING in the case of some bad XML.  I caught one of them but this is a good spot.  Is this even a current project still though?  I don't know where we are at with all this.  Toffer did something completely altered didn't he?

Reply:
Must admit I avoid setting up FAsserts because I figure I'll get myself confused on whether I should be expressing a statement that I expect to be true or false.  But this was, as mentioned, an unnecessary artifact from when this was originally a tag expressing a buildingclass rather than a direct building type reference.